### ISSUE TITLE
Docker Desktop 4.17.1 Win10 Pro/WSL2: eats up 16 GB of RAM in few minutes after start

### ISSUE THREAD

**Issue by ordnko** [docker/for-win#13325]
- [x] I have tried with the latest version of Docker Desktop
  - [ ] I have tried disabling enabled experimental features
  - [ ] I have uploaded Diagnostics
  - Diagnostics ID:

### Actual behavior
vmmem process associated with WSL2 eats up 16 GB of RAM in few minutes after Docker Desktop started
### Expected behavior
It does not consume that many memory.
### Information

  - Windows Version: Windows 10 Enterprise 22H2 19045.2728
  - Docker Desktop Version: 4.17.1
  - WSL2 or Hyper-V backend?  WSL2
  - Are you running inside a virtualized Windows e.g. on a cloud server or a VM: No

### Output of `& "C:\Program Files\Docker\Docker\resources\com.docker.diagnose.exe" check`
PS C:\Program Files\Docker\Docker\resources> ./com.docker.diagnose.exe check
[2023-03-22T16:57:20.198965500Z][com.docker.diagnose.exe][I] set path configuration to OnHost
Starting diagnostics

[PASS] DD0027: is there available disk space on the host?
[PASS] DD0028: is there available VM disk space?
[PASS] DD0002: does the bootloader have virtualization enabled?
[SKIP] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0022: is the Virtual Machine Platform Windows Feature enabled?
[PASS] DD0021: is the WSL 2 Windows Feature enabled?
[PASS] DD0024: is WSL installed?
[PASS] DD0025: are WSL distros installed?
[PASS] DD0026: is the WSL LxssManager service running?
[PASS] DD0029: is the WSL 2 Linux filesystem corrupt?
[PASS] DD0035: is the VM time synchronized?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[PASS] DD0013: is the $PATH ok?
[PASS] DD0003: is the Docker CLI working?
[PASS] DD0005: is the user in the docker-users group?
[PASS] DD0038: is the connection to Docker working?
[PASS] DD0014: are the backend processes running?
[PASS] DD0007: is the backend responding?
[PASS] DD0008: is the native API responding?
[PASS] DD0009: is the vpnkit API responding?
[PASS] DD0010: is the Docker API proxy responding?
[PASS] DD0006: is the Docker Desktop Service responding?
[SKIP] DD0030: is the image access management authorized?
[PASS] DD0033: does the host have Internet access?
[PASS] DD0002: does the bootloader have virtualization enabled?
[PASS] DD0018: does the host support virtualization?
[PASS] DD0001: is the application running?
[PASS] DD0022: is the Virtual Machine Platform Windows Feature enabled?
[PASS] DD0021: is the WSL 2 Windows Feature enabled?
[PASS] DD0024: is WSL installed?
[PASS] DD0025: are WSL distros installed?
[PASS] DD0026: is the WSL LxssManager service running?
[PASS] DD0029: is the WSL 2 Linux filesystem corrupt?
[PASS] DD0035: is the VM time synchronized?
[PASS] DD0017: can a VM be started?
[PASS] DD0016: is the LinuxKit VM running?
[PASS] DD0011: are the LinuxKit services running?
[PASS] DD0004: is the Docker engine running?
[PASS] DD0015: are the binary symlinks installed?
[PASS] DD0031: does the Docker API work?
[WARN] DD0032: do Docker networks overlap with host IPs? network bridge has subnet 172.17.0.0/16 which overlaps with host IP 172.17.160.1

Please note the following 1 warning:

1 : The check: do Docker networks overlap with host IPs?
    Produced the following warning: network bridge has subnet 172.17.0.0/16 which overlaps with host IP 172.17.160.1

If the subnet used by a Docker network overlaps with an IP used by the host, then containers
won't be able to contact the overlapping IP addresses.

Try configuring the IP address range used by networks: in your docker-compose.yml.
See https://docs.docker.com/compose/compose-file/compose-file-v2/#ipv4_address-ipv6_address

No fatal errors detected.
### Steps to reproduce the behavior

Run Docker Desktop and wait for a while. There is no need in running any containers to run into the issue.

**Comment by ehsomma** [comment]
I get the same issue after upgrade to 4.17.1 on win 10.

**Comment by thelgeson** [comment]
I am currently having this issue after installing 4.17.1 on windows 10.  I actually see a large jump in RAM usage right away, but after some containers were running for a few hours, the vmemm process slowly increased to 47 GB of memory used and exhausted my system resources.  After removing all my containers, no change in memory usage and I had to restart my machine to recover.

**Comment by markbaird** [comment]
I see this same issue. I don't even need to run any containers. I can stop all containers, stop Docker Desktop, restart WSL2 to get the RAM usage down, then I can simply start Docker Desktop and let it sit for a while with **no running containers** and after a few minutes it starts eating up all my RAM. The `Vmmem` process will start consuming about 1GB every 10 seconds or so, until my system has no more available RAM.

This is on Windows 10 Pro, WSL2, Docker Desktop 4.17.1 (101757)

I have disabled Extensions, and all experimental features.

**Comment by ehsomma** [comment]
I could stop the RAM increasing following this [workaround](https://support.itsolver.net/hc/en-au/articles/5742283229967-How-to-fix-vmmem-high-memory-usage-with-Docker-WSL2), however, with this version of docker desktop running, windows id freezing too much (impossible to use).

**Comment by a-ggghost** [comment]
Same here. Win 10 Home 22H2 19045.2788, Docker Desktop 4.17.1, WSL2
I can limit the resources used with .wslconfig, but Docker Desktop without any containers running ramps up to max allowed RAM utilization within minutes of windows login and doesn't go back down.

**Comment by thelgeson** [comment]
> WSL2 I can limit the resources used with .wslconfig, but Docker Desktop without any containers running ramps up to max allowed RAM utilization within minutes of windows login and doesn't go back down.

This is also worked for me to limit the usage, but I see the same behavior of RAM Utilization hitting the max allowed within minutes.

I downgraded to 4.17.0 (uninstall and re-install new version) and I see the same issue.... I am wondering if the upgrade did some irreparable damage to the wsl configuration?  I don't know enough to debug this further.  Anyone else try downgrading?

**Comment by markbaird** [comment]
> I downgraded to 4.17.0 (uninstall and re-install new version) and I see the same issue.... I am wondering if the upgrade did some irreparable damage to the wsl configuration? I don't know enough to debug this further. Anyone else try downgrading?

I've been seeing this behavior for a few versions now.

**Comment by Dalenir** [comment]
Also it seems much slower for me (win10) after maybe 4.16 update. Second time I trying new version.

**Comment by markbaird** [comment]
I just upgraded to version 4.18.0 and I'm still seeing the same behavior.

**Comment by bpoxy** [comment]
This has been a problem in every version after 4.15.0 (https://github.com/docker/for-win/issues/13227) but seems to get worse every release.

**Comment by junedzs** [comment]
![image](https://user-images.githubusercontent.com/92318297/232408882-2df9b3fc-a25a-47be-bb18-5438f721d853.png)

Any solution?

**Comment by RouisIbra** [comment]
Windows fixed the memory leaks of WSL2 in recent updates so why docker still have this issue?

**Comment by RouisIbra** [comment]
> Same here. Win 10 Home 22H2 19045.2788, Docker Desktop 4.17.1, WSL2 I can limit the resources used with .wslconfig, but Docker Desktop without any containers running ramps up to max allowed RAM utilization within minutes of windows login and doesn't go back down.

Setting memory limit will make WSL write to swap which writes a lot of GBs to the disk and by time it will reduce the lifespan of your hard drive. I suppose this should be a temporary workaround till Docker devs figure out why is it consuming so much RAM compared to Linux and MacOS and fix this issue.

**Comment by markbaird** [comment]
I just upgraded to version `4.19.0` and I'm still seeing the same behavior. See the attached screenshot. I have **ZERO** containers running, and Docker Desktop is taking up over 30GB of RAM a few minutes after starting it up:

 
![Screenshot 2023-05-01 101415](https://user-images.githubusercontent.com/1950062/235465299-51f33464-e26a-437c-91dd-0c19bfa48db2.png)

**Comment by ctalledo** [comment]
Hi @markbaird, thanks for adding your observations on this issue.

> I have ZERO containers running, and Docker Desktop is taking up over 30GB of RAM a few minutes after starting it up.

To help us root cause, could you provide the following info:

1) Amount of Physical RAM on your host. By default WSL will allocate up to 50% of the RAM to the WSL distros as needed, unless capped to a lower value in the `.wslconf` file.

2) Do you have any other WSL distros running (`wsl -l -v`)? (i.e., other than the `docker-desktop` and `docker-desktop-data` created by Docker Desktop). All WSL distros share the same underlying Linux kernel instance, so mem consumed by one distro will affect all other distros.

3) Without any containers running, in the Windows task manager, what's the memory consumption of the vmmemWSL process.

4) Please provide the output of `top` within the Docker-Desktop VM, so we can see what's consuming memory in there.

```
$ docker run --rm --pid=host ubuntu top -b -n 1 -o +%MEM
```

Thanks.

**Comment by ctalledo** [comment]
Thanks @markbaird for the quick response, very helpful.

Based on the output of `top` inside the WSL VM, it seems most of the mem consumption comes from the Linux kernel buffers+cache:

```
MiB Mem :  40121.9 total,   9680.5 free,   3052.0 used,  27389.4 buff/cache     <<< HERE
```

Docker Desktop processes are consuming < ~6% of the memory, which is a bit high but reasonable in your host:

```
 1377 root      20   0 3423188 490396  62408 S  26.7   1.2  13:07.90 dockerd
  805 root      20   0  801512 122816  15628 S  40.0   0.3   5:06.57 lifecycle+
  398 root      20   0  258760 108964      4 S   0.0   0.3   0:01.00 artifacto+
  988 root      20   0 1579024  61580  32368 S   0.0   0.1   5:24.54 containerd
  257 root      20   0  737636  39380  21132 S   0.0   0.1   5:31.56 containerd
  652 root      20   0  759388  25268  12452 S   0.0   0.1   0:37.03 dns-forwa+
  604 root      20   0  726304  23396  11724 S   0.0   0.1   0:04.60 diagnosti+
 1133 root      20   0  722388  20396   3040 S   0.0   0.0   0:00.24 syn-filter
  536 root      20   0  710892  17116   6068 S   0.0   0.0   0:17.64 container+
```

I suspect what's happening is that the memory consumed by the Linux kernel buffer/caches inside the WSL VM is not returned back to the host once all containers are stopped (as they ideally should).

Could you also show me the output of `top` on the Ubuntu distro:

```
C:\> wsl -d Ubuntu
$ top -b -n 1 -o +%MEM
```

I ask because the Ubuntu distro shares the kernel and VM physical mem with the `docker-desktop` and `docker-desktop-data` distros, so I want to see if there are any processes in the Ubuntu distro that could be consuming memory.

Thanks again.

**Comment by ctalledo** [comment]
@markbird: one more thing; after providing the data in the prior comment, try executing the following command inside the Ubuntu distro:

```
sudo sh -c 'echo 3 > /proc/sys/vm/drop_caches'
```

This will tell the Linux kernel to free up its page cache. This should in turn reduce the buff/cache memory consumption reported by `top`, and if things are working correctly it should cause the host memory consumed by the `vmmem*` process to drop significantly after some time (< 1 minute), meaning that the memory has been reclaimed by the host Windows OS.

**Comment by markbaird** [comment]
@ctalledo 

```
$ top -b -n 1 -o +%MEM
top - 16:46:47 up 2 days,  8:35,  0 users,  load average: 5.96, 1.92, 0.78
Tasks:  10 total,   1 running,   9 sleeping,   0 stopped,   0 zombie
%Cpu(s): 47.4 us, 28.1 sy,  0.0 ni,  3.5 id, 15.8 wa,  0.0 hi,  5.3 si,  0.0 st
MiB Mem :  40121.9 total,   9067.3 free,   5819.2 used,  25235.4 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.  33753.6 avail Mem

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
  129 root      20   0  729872  23176  12188 S   0.0   0.1   0:06.42 docker-desktop-
  143 mbaird    20   0  768564  16000    104 S   0.0   0.0   0:21.97 docker
  301 mbaird    20   0   10188   5216   3368 S   0.0   0.0   0:00.06 bash
  458 mbaird    20   0   10744   3532   3080 R   0.0   0.0   0:00.00 top
    1 root      20   0     916    524    476 S   0.0   0.0   0:00.03 init
  300 root      20   0     924    100     20 S   0.0   0.0   0:00.00 init
  299 root      20   0     924     92     20 S   0.0   0.0   0:00.00 init
  127 root      20   0     908     84     20 S   0.0   0.0   0:00.00 init
  128 root      20   0     908     84     20 S   0.0   0.0   0:00.01 init
  142 root      20   0     908     84     20 S   0.0   0.0   0:00.00 init
```

Note that if I disable Docker Desktop from starting on system boot, and restart my computer, I can use WSL Ubuntu without any runaway memory usage happening. It's only after I start Docker Desktop that it starts eating up all my RAM.

Also, I saw the `sudo sh -c 'echo 3 > /proc/sys/vm/drop_caches'` trick several weeks ago from a comment on the related issue [here](https://github.com/microsoft/WSL/issues/8725). Running that will get the memory usage to go down *temporarily*. After a few minutes the memory usage starts going back up and I have to run the command again. If I let the memory usage get up to the maximum 40GB I have allocated to WSL, then all my docker containers and docker build commands start running extremely slow until I do the `drop_caches` trick again. Sometimes after I run that command a lot, maybe 20 times in an hour, it stops eating up more memory and may actually work as expected for a while.

**Comment by ctalledo** [comment]
Thanks again @markbaird.

>  It's only after I start Docker Desktop that it starts eating up all my RAM.

Got it. Are you just running containers, are also building container images? I ask because the latter is generally more disk intensive and therefore would result in higher kernel page cache usage. 

> Also, I saw the sudo sh -c 'echo 3 > /proc/sys/vm/drop_caches' trick several weeks ago from a comment on the related issue https://github.com/microsoft/WSL/issues/8725. Running that will get the memory usage to go down temporarily

Yes I am not suggesting it's a true fix, just thought I would mention it as a temporary work-around. 

We (Docker Inc) probably need to improve Docker Desktop to automatically release the page cache mem inside the WSL VM when no containers are running. The problem is that other WSL distros (e.g., your Ubuntu distro) may be using some portion of that cache, since all WSL distros on the host share the same kernel.

**Comment by markbaird** [comment]
@ctalledo I'm doing software development and testing on this host, which involves lots of docker builds in addition to running containers.

**Comment by ctalledo** [comment]
FYI, another way to decrease the memory consumption inside the WSL VM, particularly memory consumed by the Linux kernel caches (e.g., page cache) is to remove unused Docker container images (e.g., `docker image rm ...`). When those are removed, the Linux kernel frees page cache memory associated with those images.

**Comment by RouisIbra** [comment]
> FYI, another way to decrease the memory consumption inside the WSL VM, particularly memory consumed by the Linux kernel caches (e.g., page cache), is to remove unused Docker container images (e.g., `docker image rm ...`). When those are removed, the Linux kernel frees page cache memory associated with those images.

I believe page cache in memory was made to improve performance for systems with HDD storage, right? Can you provide an option to store it in disk?

**Comment by markbaird** [comment]
> FYI, another way to decrease the memory consumption inside the WSL VM, particularly memory consumed by the Linux kernel caches (e.g., page cache), is to remove unused Docker container images (e.g., `docker image rm ...`). When those are removed, the Linux kernel frees page cache memory associated with those images. Once that memory is free, [WSL will reclaim them and give them back to the host's Windows OS.](https://devblogs.microsoft.com/commandline/memory-reclaim-in-the-windows-subsystem-for-linux-2/).

@ctalledo I had actually noticed that it seemed to get somewhat better if I kept my docker images pruned down. I didn't mention it because I had tried so many different things that I wasn't sure that was what made the difference. Unfortunately keeping the images pruned isn't easy to do when I am building new docker images often. And I also need to keep some fairly large images on my host all the time.

This memory usage problem wasn't always the case though. I've gone years now with having many gigabytes of images on my computer without Docker Desktop going crazy and sucking up all my RAM on startup. It's only fairly recently, perhaps the last 6 months or so, that Docker Desktop has suddenly been so proactive about this page caching. I definitely feel like this is new behavior that hasn't always been there. I've been using Docker Desktop for Windows since it was first in beta, so I've tested every iteration of it, and only fairly recently has it started being so hungry for memory in my experience.

**Comment by RouisIbra** [comment]
> @RouisIbra The whole point is that it is taking something on disk and caching it in RAM. If you moved the cache to disk it would just be making a duplicate copy of something already on disk. At that point you would be better off just disabling the cache all together.

Can you make docker read `page cache` directly from disk without copying it to RAM?

**Comment by bpoxy** [comment]
> FYI, another way to decrease the memory consumption inside the WSL VM, particularly memory consumed by the Linux kernel caches (e.g., page cache), is to remove unused Docker container images (e.g., `docker image rm ...`). When those are removed, the Linux kernel frees page cache memory associated with those images. Once that memory is free, [WSL will reclaim them and give them back to the host's Windows OS.](https://devblogs.microsoft.com/commandline/memory-reclaim-in-the-windows-subsystem-for-linux-2/).

I experienced the reported issue with only one image on my system so I'm not sure that the number of images is relevant.

**Comment by ctalledo** [comment]
Hi @markbaird, 

> This memory usage problem wasn't always the case though. I've gone years now with having many gigabytes of images on my computer without Docker Desktop going crazy and sucking up all my RAM on startup. It's only fairly recently, perhaps the last 6 months or so, that Docker Desktop has suddenly been so proactive about this page caching.

That's good info, thanks. I'll revert to older Docker Desktop versions to see what's going on. It's a bit tricky because Docker Desktop does not control the page caching, it's the WSL Linux kernel that does it in an effort to improve performance inside the VM. But maybe recent versions of Docker Desktop are doing more I/O inside the VM, causing the page cache to increase. I'll take a look.

Rest assured we are working on fixing this so that if Docker Desktop is idle, memory consumed by the DD VM gets returned back to the host Windows OS as it should.

**Comment by ctalledo** [comment]
Hi @RouisIbra,

> Can they make docker read page cache directly from disk without copying it to RAM?

There may be a way inside the Linux kernel to disable the kernel's page cache altogether, but that would likely reduce performance significantly (e.g., docker builds would run much slower). The page cache has a very good purpose, so it's good to have it enabled. We just want to ensure that when it's no longer needed, the memory it consumes can go back to the host Windows OS.

**Comment by ctalledo** [comment]
Hi @bpoxy,

> I experienced the reported issue with only one image on my system so I'm not sure that the number of images is relevant.

The size of the one image could also make a difference (e.g., if it's made up of many files or layers for example). In general, the more disk IO is done inside the Docker Desktop VM, the larger the page cache inside of it.

**Comment by samuk190** [comment]
anyone confirm using earlier version (4.15) does not have any memory leak at all ?

**Comment by nogenem** [comment]
> anyone confirm using earlier version (4.15) does not have any memory leak at all ?

I'm using 4.14.1 for a couple of weeks and the problem still happens, maybe not as much, but... ;/

**Comment by ctalledo** [comment]
I did a comparison between DD 4.15 and 4.19 for memory usage and don't see any large differences. Please see my response here:  https://github.com/docker/for-win/issues/13227#issuecomment-1558534914
